editoast/src/views/v2/timetable/stdcm.rs

```diff
@@ -233,6 +246,87 @@ async fn stdcm(
     Ok(Json(stdcm_response))
 }
 
+/// get the maximum run time, compute it if unspecified.
+/// returns an enum with either the result or a SimulationResponse if it failed
+async fn get_max_run_time(
+    db_pool: Arc<DbConnectionPool>,
+    redis_client: Arc<RedisClient>,
+    core_client: Arc<CoreClient>,
+    data: &STDCMRequestPayload,
+    infra: &Infra,
+    rolling_stock: &RollingStockModel,
+) -> Result<MaxRunningTimeResult> {
+    if data.maximum_run_time.is_some() {
+        return Ok(MaxRunningTimeResult::MaxRunningTime {
+            value: data.maximum_run_time.unwrap(),
+        });
+    }
+
+    let train_schedule = TrainSchedule {
+        id: 0,
+        train_name: "".to_string(),
+        labels: vec![],
+        rolling_stock_name: rolling_stock.name.clone(),
+        timetable_id: 0, // TODO: do we actually need a valid ID?
+        start_time: data.start_time,
+        schedule: vec![],
+        margins: build_single_margin(data.margin),
+        initial_speed: 0.0,
+        comfort: data.comfort,
+        path: convert_steps(&data.steps),
+        constraint_distribution: Default::default(),
+        speed_limit_tag: data.speed_limit_tags.clone(),
+        power_restrictions: vec![],
+        options: Default::default(),
+    };
+
+    // TODO: check that the result isn't actually saved in db
+    let sim_result =
+        train_simulation(db_pool, redis_client, core_client, &train_schedule, infra).await?;
+
+    let total_stop_time: u64 = data
+        .steps
+        .iter()
+        .map(|step: &PathfindingItem| step.duration.unwrap_or_default())
+        .sum();
+
+    return Ok(match sim_result {
+        SimulationResponse::Success { provisional, .. } => MaxRunningTimeResult::MaxRunningTime {
+            value: *provisional.times.last().unwrap() * 2 + total_stop_time,
+        },
+        err => MaxRunningTimeResult::Error {
+            error: Box::from(err),
+        },
+    });
+}
+
+/// Convert the list of pathfinding items into a list of path item
+fn convert_steps(steps: &[PathfindingItem]) -> Vec<PathItem> {
+    return steps
+        .iter()
+        .map(|step| PathItem {
+            id: Default::default(),
+            deleted: false,
+            location: step.location.clone(),
+        })
+        .collect();
+}
+
+/// Build a margins object with one margin value covering the entire range
+fn build_single_margin(margin: Option<MarginValue>) -> Margins {
+    match margin {
+        None => Margins {
+            boundaries: vec![],
+            values: vec![],
+        },
+        Some(m) => Margins {
+            boundaries: vec![NonBlankString("0".into()), NonBlankString("-1".into())],
```

Comment:
`NonBlankString` implements `From<&str>` so you should be able to do the following.

```suggestion
            boundaries: vec![NonBlankString::from("0"), NonBlankString:from("-1")],
```

Reply:
Thanks for the tip

What's weird is that the `boundaries` vec *should* be empty, I was wrong on that first version. It's empty when I `git show` and the branch is up-to-date on the remote, but github still shows the old version :thinking: It seems to struggle with processing the last push

Reply:
Hmm, I did notice some weird spinner at the top of the PR when I was reviewing. I don't know what it was. Since it's an outdated comment, I'm resolving.